src/run.rs

```diff
@@ -273,44 +273,37 @@ pub fn run() {
       );
     }
   } else {
-    let name;
-    'outer: loop {
-      for candidate in &["justfile", "Justfile"] {
-        match fs::metadata(candidate) {
-          Ok(metadata) => {
-            if metadata.is_file() {
-              name = *candidate;
-              break 'outer;
-            }
-          }
-          Err(error) => {
-            if error.kind() != io::ErrorKind::NotFound {
-              die!("Error fetching justfile metadata: {}", error)
-            }
-          }
+    match search(
+      env::current_dir()
+        .as_ref()
+        .expect("Error getting current directory"),
+    ) {
+      Ok(name) => {
+        if matches.is_present("EDIT") {
+          edit(name);
         }
+        text = fs::read_to_string(name)
+          .unwrap_or_else(|error| die!("Error reading justfile: {}", error));
       }
-
-      match env::current_dir() {
-        Ok(pathbuf) => {
-          if pathbuf.as_os_str() == "/" {
-            die!("No justfile found.");
-          }
+      Err(err) => match err {
+        SearchError::NotFound => {
+          die!("No justfile found");
         }
-        Err(error) => die!("Error getting current dir: {}", error),
-      }
-
-      if let Err(error) = env::set_current_dir("..") {
-        die!("Error changing directory: {}", error);
-      }
-    }
-
-    if matches.is_present("EDIT") {
-      edit(name);
+        SearchError::MultipleCandidates { candidates } => {
```

Comment:
In the interest of keep the size of the run function down, let's move `SearchError` into its own file, `src/search_error.rs`, and add an `impl Display for SearchError {...}`, then all the error formatting can be done in the `impl` block.

Then, in main, we can do:

```
let justfile = match search(&current_dir) {
  Ok(justfile) => justfile,
  Err(search_error) => die!("{}", search_error),
};
```